src/hotspot/share/runtime/arguments.cpp

```diff
@@ -744,20 +745,83 @@ bool Arguments::verify_special_jvm_flags(bool check_globals) {
 }
 #endif
 
-// Parses a size specification string.
-bool Arguments::atojulong(const char *s, julong* result) {
-  julong n = 0;
+template <typename T, ENABLE_IF(std::is_signed<T>::value), ENABLE_IF(sizeof(T) == 4)> // signed 32-bit
+static bool parse_integer_impl(const char *s, char **endptr, int base, T* result) {
+  STATIC_ASSERT(sizeof(long) >= 4); // need to have enough bits
+  errno = 0; // errno is thread safe
+  long v = strtol(s, endptr, base);
+  if (errno != 0 LP64_ONLY(|| v < min_jint || v > max_jint)) {
+    // long is 64-bit on LP64, so we need explicit range check.
+    return false;
+  }
+  *result = static_cast<T>(v);
+  return true;
+}
 
-  // First char must be a digit. Don't allow negative numbers or leading spaces.
-  if (!isdigit(*s)) {
+template <typename T, ENABLE_IF(!std::is_signed<T>::value), ENABLE_IF(sizeof(T) == 4)> // unsigned 32-bit
+static bool parse_integer_impl(const char *s, char **endptr, int base, T* result) {
+  if (s[0] == '-') {
     return false;
   }
+  STATIC_ASSERT(sizeof(unsigned long) >= 4); // need to have enough bits
+  errno = 0; // errno is thread safe
+  unsigned long v = strtoul(s, endptr, base);
+  if (errno != 0 LP64_ONLY(|| v > max_juint)) {
+    // unsigned long is 64-bit on LP64, so we need explicit range check.
```

Comment:
Same issue here about Windows `long` being 32 bits.

Reply:
Hi Kim, thanks for the comments. I have reverted the code to use strtoll/strtoull as in 2a5e3e6df3e61b15126ca0582aada57955a25a07, and updated the comments to explain why.